### DISCUSSION TITLE
Version 3

### DISCUSSION THREAD

**Discussion by sergiodxa** [sergiodxa/remix-auth#94] [Ideas]
This discussion will work as an umbrella for all the things I want to do for v3 of Remix Auth.

## Multi Package

Right now, Remix Auth is a single package with everything built-in, this caused a few issues where KCDStrategy was importing the crypto Node.js module making the whole library impossible to use in Cloudflare, even if you didn't used such strategy.

The idea is to split the Remix Auth package is multiple packages.

1. `remix-auth` will include the Authenticator, Authorizer and the Strategy interface
2. `remix-auth-oauth2` will include the OAuth2Strategy
3. `remix-auth-password` will include the `LocalStrategy` (renamed to `PasswordStrategy`)
5. `remix-auth-auth0` will include the `Auth0Strategy`
6. `remix-auth-github` will include the GitHubStrategy
7. `remix-auth-google` will include the GoogleStrategy
8. `remix-auth-kcd` will include the KCDStrategy
9. And in a similar way the rest of the strategies will become their own `remix-auth-*` packages.

This will allow you to install only the packages you need, so you can install `remix-auth` and `remix-auth-github` for example, and ignore the rest, or `remix-auth-credentials`.

It will also make easier to build strategies that depends on Node.js built-in modules or Cloudflare APIs or in the future Deno APIs.

It would also mean the community has an example of how to publish their own strategies, and I can focus on a few packages (the code, the oauth2, credentials, auth0 and kcd probably)

## Better Errors

Right now Remix Auth set the errors in the session and redirect or throw a response.

It should support throwing actual errors or return then.

**Comment by edwin177** [comment]
It looks like v3 has already been published, but `remix-auth-password` has not been?

**Reply by sergiodxa** [reply]
I'm still working on that, I will do a `remix-auth-form` strategy, it will give you the FormData in the verify callback, you can get the username or email and password from there, or anything else you may need, so it's more open to extra fields.

**Reply by sergiodxa** [reply]
You can see the repo in https://github.com/sergiodxa/remix-auth-form (I'm working on pushing the v1)
